### PR TITLE
[DTensor] Add DTensor experimental op for LayerNorm backward sharding rule propogation

### PR DESCRIPTION
**Pull request by yoyoyocmu**
Summary: This diff is only for prototype to unblock the TP work. PyTorch distributed team is working on a more generic backward op for `aten.layer_norm`. Will remove this op from the experimental file once it is ready.

Test Plan:
**Local Test**:
Accuracy:
- Dtensor + Checkpoint: first run loss: P884569822 (on-par with baseline: P884213363)
- 2nd by loading saved checkpoint: P884583429 (on-par with baseline: P884271869)

Trace:
- Collective functions are inserted automatically.
- Example: https://fburl.com/perfdoctor/l567ww1x

**MAST Test**:
With: trainer = 128, batch_size=512
- NE on-par:
(see: 4441_ep_bs512_2fsdp_tp_sp_dtensor)
 {F1155318138}

Differential Revision: D51490868

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @kiukchung @lucasllc

### CODE REVIEW

File: torch/distributed/_tensor/ops/experimental_ops.py [hunk 1]
```diff
@@ -88,3 +93,51 @@ def nll_loss_backward_rules(op_schema: OpSchema) -> OutputSharding:
...
+        assert isinstance(weight, DTensorSpec)
+        assert all(isinstance(s, Replicate) for s in weight.placements)
+    if bias is not None:
+        assert isinstance(bias, DTensorSpec)
+        assert all(isinstance(s, Replicate) for s in bias.placements)
+
+    weight_grad = (
+        DTensorSpec(
+            mesh=weight.mesh,
+            placements=tuple([_Partial()] * weight.mesh.ndim),
```
Comment (wanchaol) [inline torch/distributed/_tensor/ops/experimental_ops.py:122]: curious how are you triggering the backward allreduce for the gradients now? since the param.grads are partial tensors, it's by default lazy staying as partial dtensor, if we use dtensor alone the allreduce will automatically be triggered in the optimizer.step. But if we combine with FlatParam FSDP together we would need to eagerly trigger this once the gradient is ready (i.e. via register_hook on parameters)

**Review by wanchaol — APPROVED** [review]
stamp to unblock as it's a experimental op

File: torch/distributed/_tensor/ops/experimental_ops.py (hunk 1 above)
Comment (yoyoyocmu) [reply torch/distributed/_tensor/ops/experimental_ops.py:122]: Ideally we don't want to overwrite the placement always as Partial() or any "hardcoded" placements, instead, we should rely on the sharing propagation rule to make the decision. Specifically for the DTensor-based experimental WHEN model, the "overwrite" operation is unnecessary. However, to maintain consistency with the current behavior here: https://github.com/pytorch/pytorch/blob/main/torch/distributed/_spmd/experimental_ops.py#L290 I introduced these lines as well.

If we have concerns, I can remove them.
